### DISCUSSION TITLE
Closing the connection (async-nats)

### DISCUSSION THREAD

**Discussion by mnordstr** [nats-io/nats.rs#1335] [Q&A]
Hi,

I'm trying to reconnect after the JWT token has expired, I'm signaling my app from the event_callback when I detect an authentication problem to drop everything and start again. This basically works but the old async nats client keeps on trying. I can not seem to kill the connection, even drop(client) does not help. Can't seem to find anything in the documentation regarding closing connections.

**Comment by Jarema** [comment]
Hey.

First of all, you should not need to shut everything down if you use [auth callback](https://docs.rs/async-nats/0.37.0/async_nats/struct.ConnectOptions.html#method.with_auth_callback)

Regarding dropping the connection - connection will be dropped when all `Client` instances are dropped.
Next release will also contain `drain` functionality which will be alternative way to achieve the same goal.

**Reply by mnordstr** [reply]
Thanks for the insights. I'm pretty sure I have dropped all Clients but maybe I'm leaking somewhere.

Trying to tackle it using the auth callback. Is the nonce that auth_callback gets the same as when using with_jwt? Cleaned up code below but it gives me nats: authorization violation:

```rust
with_auth_callback(move |nonce| async move {
    println!("Got JWT: {:?}", jwt);
    println!("Got Seed: {:?}", seed);

    let key_pair = Arc::new(nkeys::KeyPair::from_seed(seed.as_str()).unwrap());

    let mut auth = async_nats::Auth::new();
    auth.jwt = Some(jwt.clone());
    println!("Nonce: {:?}", nonce);
    let sign = key_pair.sign(&nonce).map_err(async_nats::AuthError::new)?;
    auth.signature = Some(URL_SAFE_NO_PAD.encode(sign));
    println!("Sig: {:?}", auth.signature);

    Ok(auth)
}
```

**Reply by Jarema** [reply]
try not encoding the signature as we have this in code:

```rust
 connect_info.signature = auth
  .signature
  .map(|signature| URL_SAFE_NO_PAD.encode(signature));
```

**Reply by mnordstr** [reply]
It expects a String as the signature, if I convert the `Vec<u8>` result from the sign function to a String using String::from_utf8 it will fail with FromUtf8Error.

```rust
let sign = key_pair.sign(&nonce).map_err(async_nats::AuthError::new)?;
auth.signature = Some(String::from_utf8(sign).unwrap());
```

**Reply by mnordstr** [reply]
I tried to change the signature in the Auth struct of the async-nats crate to a `Vec<u8>` and after that everything works nicely. I have a feeling the String type there is not correct?

Do you want a real issue + PR?

**Reply by mnordstr** [reply]
Added a PR for this, #1337
